File: CStoreManagement/app/src/main/java/com/c_store/zhiyazhang/cstoremanagement/presenter/contract/ContractAdapter.java
package com.c_store.zhiyazhang.cstoremanagement.presenter.contract;

import android.annotation.SuppressLint;
import android.content.Context;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.CardView;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.c_store.zhiyazhang.cstoremanagement.R;
import com.c_store.zhiyazhang.cstoremanagement.bean.ContractBean;
import com.c_store.zhiyazhang.cstoremanagement.bean.ContractResult;
import com.c_store.zhiyazhang.cstoremanagement.utils.onclick.RecyclerOnItemClickListener;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Created by zhiya.zhang
 * on 2017/5/10 14:58.
 */

public class ContractAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    private ContractResult cr;

    private List<ContractBean> cbs;
    private RecyclerOnItemClickListener onClick;
    private Context context;
    private boolean isEdit;
    private static final int TYPE_ITEM = 0;  //普通Item View
    private static final int TYPE_FOOTER = 1;  //顶部FootView
    private int load_more_status = 0;//上拉加载更多状态-默认为0
    static final int PULLUP_LOAD_MORE = 0;//上拉加载更多
    public static final int LOADING_MORE = 1;//正在加载中


    void setOnItemClickLitener(RecyclerOnItemClickListener recyclerOnItemClickListener) {
        this.onClick = recyclerOnItemClickListener;
    }

    public void setIsEdit(boolean isEdit) {
        this.isEdit = isEdit;
    }

    ContractAdapter(ContractResult cr, Context context) {
        this.cr = cr;
        this.cbs = cr.getDetail();
        this.context = context;
    }

    public List<ContractBean> getCbs() {
        return cbs;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        if (viewType == TYPE_ITEM) {
            return new ViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.commodify_item, parent, false));
        } else if (viewType == TYPE_FOOTER) {
            return new FooterViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.footview, parent, false));
        }
        return null;
    }

    @SuppressLint("ClickableViewAccessibility")
    @Override
    public void onBindViewHolder(final RecyclerView.ViewHolder holder, int position) {
        if (holder instanceof ViewHolder) {
            ((ViewHolder) holder).cid.setText(cbs.get(position).getcId());
            ((ViewHolder) holder).cname.setText(cbs.get(position).getcName());
            ((ViewHolder) holder).cin.setText(Integer.toString(cbs.get(position).getInventory()));
            ((ViewHolder) holder).cnc.setText(Integer.toString(cbs.get(position).getTonightCount()));
            ((ViewHolder) holder).cprice.setText(Double.toString(cbs.get(position).getcPrice()));
            ((ViewHolder) holder).editCdc.setText(Integer.toString(cbs.get(position).getTodayCount()));
            if (cbs.get(position).getTodayStore() != 0) {
                ((ViewHolder) holder).card.setBackgroundColor(ContextCompat.getColor(context, R.color.blue));
            } else {
                ((ViewHolder) holder).card.setBackgroundColor(ContextCompat.getColor(context, R.color.bg_color));
            }
            Glide.with(context).load(cbs.get(position).getImg_url()).placeholder(R.mipmap.loading).error(R.mipmap.load_error).crossFade().centerCrop().into(((ViewHolder) holder).img);
            if (onClick != null) {
                //不知道为什么明明设置了限制，但是在编辑模式下点击card还是会执行监听事件，只能通过强行enabled强行禁止掉
                if (isEdit) {
                    ((ViewHolder) holder).add.setOnTouchListener(new View.OnTouchListener() {
                        @Override
                        public boolean onTouch(View v, MotionEvent event) {
                            onClick.onTouchAddListener(cbs.get(holder.getLayoutPosition()), holder.getLayoutPosition(), event);
                            return true;
                        }
                    });
                    ((ViewHolder) holder).less.setOnTouchListener(new View.OnTouchListener() {
                        @Override
                        public boolean onTouch(View v, MotionEvent event) {
                            onClick.onTouchLessListener(cbs.get(holder.getLayoutPosition()), holder.getLayoutPosition(), event);
                            return true;
                        }
                    });
                    ((ViewHolder) holder).card.setEnabled(false);
                } else {
                    ((ViewHolder) holder).card.setEnabled(true);
                    ((ViewHolder) holder).card.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            onClick.onItemClick(holder.itemView, holder.getLayoutPosition());
                        }
                    });
                }
            }
            /*if (isEdit) {
                ((ViewHolder) holder).img.setVisibility(View.GONE);
            } else {
                ((ViewHolder) holder).img.setVisibility(View.VISIBLE);
            }*/
        } else if (holder instanceof FooterViewHolder) {
            FooterViewHolder fHolder = (FooterViewHolder) holder;
            switch (load_more_status) {
                case PULLUP_LOAD_MORE:
                    fHolder.footLoading.setText(context.getResources().getString(R.string.loadingUp));
                    break;
                case LOADING_MORE:
                    fHolder.footLoading.setText(context.getResources().getString(R.string.loading));
                    break;
            }
        }
    }

    @Override
    public int getItemViewType(int position) {
        if (position + 1 == getItemCount()) {
            if (cr.getTotal() > 10 && cbs.size() < cr.getTotal()) {
                return TYPE_FOOTER;
            } else {
                return TYPE_ITEM;
            }
        } else {
            return TYPE_ITEM;
        }
    }

    @Override
    public int getItemCount() {
        //当前页不足10且已有的数据少于总数据才会添加底部加载
        if (cr.getTotal() > 10 && cbs.size() < cr.getTotal()) {
            return cbs.size() + 1;
        }
        return cbs.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.my_commodify)
        public CardView card;
        @BindView(R.id.commodify_img)
        ImageView img;
        @BindView(R.id.cid)
        TextView cid;
        @BindView(R.id.cname)
        TextView cname;
        @BindView(R.id.cprice)
        TextView cprice;
        @BindView(R.id.cin)
        TextView cin;
        @BindView(R.id.cnc)
        TextView cnc;
        @BindView(R.id.add)
        public ImageButton add;
        @BindView(R.id.less)
        public ImageButton less;
        @BindView(R.id.edit_cdc)
        public TextView editCdc;

        ViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }

    //底部加载
    class FooterViewHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.foot)
        LinearLayout foot;
        @BindView(R.id.footLoading)
        TextView footLoading;
        @BindView(R.id.footProgress)
        ProgressBar footProgress;

        public FooterViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }

    void addItem(List<ContractBean> newDatas) {
        cbs.addAll(newDatas);
        notifyDataSetChanged();
    }

    /**
     * //上拉加载更多
     * PULLUP_LOAD_MORE=0;
     * //正在加载中
     * LOADING_MORE=1;
     * //加载完成已经没有更多数据了
     * NO_MORE_DATA=2;
     */
    public void changeMoreStatus(int statud) {
        load_more_status = statud;
        notifyDataSetChanged();
    }
}

File: CStoreManagement/app/src/main/java/com/cstore/zhiyazhang/cstoremanagement/presenter/contract/ContractAdapter.kt
package com.cstore.zhiyazhang.cstoremanagement.presenter.contract

import android.annotation.SuppressLint
import android.content.Context
import android.support.v4.content.ContextCompat
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import com.bumptech.glide.Glide
import com.cstore.zhiyazhang.cstoremanagement.R
import com.cstore.zhiyazhang.cstoremanagement.bean.ContractBean
import com.cstore.zhiyazhang.cstoremanagement.bean.ContractResult
import com.cstore.zhiyazhang.cstoremanagement.utils.recycler.RecyclerOnTouch
import kotlinx.android.synthetic.main.item_loading.view.*

/**
 * Created by zhiya.zhang
 * on 2017/6/14 15:58.
 */
class ContractAdapter(val cr: ContractResult, val context: Context, val onTouch: RecyclerOnTouch, var isJustLook: Boolean) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    companion object {
        private val TYPE_ITEM = 0  //普通Item View
        private val TYPE_FOOTER = 1  //底部FootView
        private var load_more_status = 0//上拉加载更多状态-默认为0
        val PULLUP_LOAD_MORE = 0//上拉加载更多
        val LOADING_MORE = 1//正在加载中
    }


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder =
            when (viewType) {//是什么页面就添加对应的
                TYPE_ITEM -> {
                    ViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_contract, parent, false))
                }
                TYPE_FOOTER -> {
                    FooterViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_loading, parent, false))
                }
                else ->
                    ViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_contract, parent, false))
            }

    @SuppressLint("ClickableViewAccessibility")
    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        when (holder) {
            is ViewHolder -> {
                holder.cid.text = cr.detail[position].cId
                holder.cname.text = cr.detail[position].cName
                holder.cin.text = cr.detail[position].inventory.toString()
                holder.cnc.text = cr.detail[position].tonightCount.toString()
                holder.cprice.text = cr.detail[position].cPrice.toString()
                holder.editCdc.text = cr.detail[position].todayCount.toString()
                holder.arriveDate.text = when (cr.detail[position].arriveDate) {
                    1 -> {
                        holder.arriveDate.setTextColor(ContextCompat.getColor(context, R.color.cstore_red))
                        context.getString(R.string.todayCount)
                    }
                    2 -> {
                        holder.arriveDate.setTextColor(ContextCompat.getColor(context, R.color.cstore_red))
                        context.getString(R.string.tomorrow_count)
                    }
                    3 -> {
                        holder.arriveDate.setTextColor(ContextCompat.getColor(context, R.color.add_less))
                        context.getString(R.string.acquired_count)
                    }
                    else -> {
                        holder.arriveDate.setTextColor(ContextCompat.getColor(context, R.color.cstore_red))
                        context.getString(R.string.tomorrow_count)
                    }
                }
                //通过判断商店修改决定是否改色
                when {
                    cr.detail[position].todayStore > 0 -> holder.myCommodify.setBackgroundColor(ContextCompat.getColor(context, R.color.add_bg))
                    cr.detail[position].todayStore < 0 -> holder.myCommodify.setBackgroundColor(ContextCompat.getColor(context, R.color.less_bg))
                    cr.detail[position].todayStore == 0 -> holder.myCommodify.setBackgroundColor(ContextCompat.getColor(context, R.color.white))
                }
                Glide.with(context).load(cr.detail[position].img_url)
                        .placeholder(R.mipmap.loading)
                        .error(R.mipmap.load_error)
                        .crossFade()
                        .into(holder.commodifyImg)
                holder.commodifyImg.setOnClickListener {
                    onTouch.onClickImage(cr.detail[position], position)
                }
                if (isJustLook) {
                    holder.alLine.visibility = View.GONE
                    holder.csCount.text = cr.detail[position].cSug.toString()
                    holder.cType.text = cr.detail[position].cType
                    holder.add.visibility = View.GONE
                    holder.less.visibility = View.GONE
                } else {
                    //提示不用管，只是说touch覆盖了辅助提示
                    holder.add.setOnTouchListener { _, event ->
                        onTouch.onTouchAddListener(cr.detail[position], event, position)
                        true
                    }
                    holder.less.setOnTouchListener { _, event ->
                        onTouch.onTouchLessListener(cr.detail[position], event, position)
                        true
                    }
                }
            }
            is FooterViewHolder -> {
                holder.bind()
            }
            else -> {
            }
        }
    }

    override fun getItemViewType(position: Int): Int {
        //确认是添加正常view还是底部加载view
        return if (position + 1 == itemCount) {
            if (cr.total > 10 && cr.detail.size < cr.total) {
                TYPE_FOOTER
            } else {
                TYPE_ITEM
            }
        } else {
            TYPE_ITEM
        }
    }

    //当前页不足10且已有的数据少于总数据才会添加底部加载
    override fun getItemCount(): Int =
            if (cr.total > 10 && cr.detail.size < cr.total) cr.detail.size + 1
            else cr.detail.size


    fun addItem(newDates: List<ContractBean>) {
        (cr.detail as ArrayList<ContractBean>).addAll(newDates)
    }

    /**
     * 上拉加载更多 PULLUP_LOAD_MORE=0
     * 正在加载中 LOADING_MORE=1
     * 加载完成已经没有更多数据了 NO_MORE_DATA=2;
     */
    fun changeMoreStatus(statud: Int) {
        load_more_status = statud
        notifyDataSetChanged()
    }


    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val cid = itemView.findViewById<TextView>(R.id.cid)!!
        val cname = itemView.findViewById<TextView>(R.id.cname)!!
        val cin = itemView.findViewById<TextView>(R.id.cin)!!
        val cnc = itemView.findViewById<TextView>(R.id.cnc)!!
        val cprice = itemView.findViewById<TextView>(R.id.cprice)!!
        val myCommodify = itemView.findViewById<LinearLayout>(R.id.my_commodify)!!
        val commodifyImg = itemView.findViewById<ImageView>(R.id.commodify_img)!!
        val less = itemView.findViewById<ImageButton>(R.id.less)!!
        val add = itemView.findViewById<ImageButton>(R.id.add)!!
        val editCdc = itemView.findViewById<TextView>(R.id.edit_cdc)!!
        val arriveDate = itemView.findViewById<TextView>(R.id.arriveDate)!!
        val cType = itemView.findViewById<TextView>(R.id.ctype)!!
        val csCount = itemView.findViewById<TextView>(R.id.cs_count)!!
        val alLine = itemView.findViewById<View>(R.id.al_line)!!
    }

    private inner class FooterViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        fun bind() = with(itemView) {
            when (load_more_status) {//通过不同状态修改显示内容
                PULLUP_LOAD_MORE -> footLoading.text = context.getString(R.string.loadingUp)
                LOADING_MORE -> footLoading.text = context.getString(R.string.loading)
            }
        }
    }
}